File: apps/server/src/execution_v2/core/execution_engine.py
"""Simplified main execution engine for diagram processing."""

import asyncio
import uuid
from typing import Dict, List, Optional, Tuple, Any
from collections import deque

from .execution_context import ExecutionContext
from .skip_manager import SkipManager
from .loop_controller import LoopController
from ..executors.base_executor import BaseExecutor
from ..executors.person_job_executor import PersonJobExecutor
from ..executors.condition_executor import ConditionExecutor
from ..executors.start_executor import StartExecutor


class ExecutionEngine:
    """Simplified main execution engine that orchestrates diagram execution."""
    
    def __init__(
        self,
        diagram: dict,
        memory_service,
        llm_service,
        streaming_update_callback=None
    ):
        """Initialize the execution engine.
        
        Args:
            diagram: The diagram configuration containing nodes and arrows
            memory_service: Service for memory management
            llm_service: Service for LLM interactions
            streaming_update_callback: Optional callback for streaming updates
        """
        self.diagram = diagram
        self.memory_service = memory_service
        self.llm_service = llm_service
        self.streaming_update_callback = streaming_update_callback
        
        # Initialize components
        self.context = ExecutionContext(execution_id=str(uuid.uuid4()))
        self.skip_manager = SkipManager()
        self.loop_controller = LoopController()
        
        # Create executors
        self.executors = self._create_executors()
        
        # Build node and arrow lookups
        self.nodes = {node['id']: node for node in diagram.get('nodes', [])}
        self.arrows = diagram.get('arrows', [])
        self._build_connections()
    
    def _create_executors(self) -> Dict[str, BaseExecutor]:
        """Create executor instances for each node type.
        
        Returns:
            Dictionary mapping node type to executor instance
        """
        return {
            'personJobNode': PersonJobExecutor(self.llm_service, self.memory_service),
            'conditionNode': ConditionExecutor(),
            'startNode': StartExecutor(),
            # Add more executors as they are implemented
        }
    
    def _build_connections(self) -> None:
        """Build connection mappings for efficient traversal."""
        self.outgoing_arrows: Dict[str, List[dict]] = {}
        self.incoming_arrows: Dict[str, List[dict]] = {}
        
        for arrow in self.arrows:
            source = arrow['source']
            target = arrow['target']
            
            if source not in self.outgoing_arrows:
                self.outgoing_arrows[source] = []
            self.outgoing_arrows[source].append(arrow)
            
            if target not in self.incoming_arrows:
                self.incoming_arrows[target] = []
            self.incoming_arrows[target].append(arrow)
    
    async def execute(self, initial_inputs: Optional[Dict[str, Any]] = None) -> Tuple[Dict[str, Any], float]:
        """Main execution loop - simplified and linear.
        
        Args:
            initial_inputs: Optional initial inputs for start nodes
            
        Returns:
            Tuple of (node_outputs, total_cost)
        """
        # Initialize execution queue with start nodes
        execution_queue = deque(self._get_start_nodes())
        
        while execution_queue:
            node_id = execution_queue.popleft()
            
            if node_id not in self.nodes:
                continue
            
            node = self.nodes[node_id]
            
            # Send streaming update
            await self._send_update(node_id, "running")
            
            # Check if we should skip
            if self._should_skip_node(node):
                self.context.set_node_output(node_id, {"skipped": True})
                await self._send_update(node_id, "skipped")
                execution_queue.extend(self._get_next_nodes(node_id))
                continue
            
            # Execute the node
            executor = self.executors.get(node['type'])
            if not executor:
                await self._send_update(node_id, "error", "No executor for node type")
                continue
            
            try:
                # Validate inputs
                error = await executor.validate_inputs(node, self.context)
                if error:
                    raise ValueError(error)
                
                # Execute
                output, cost = await executor.execute(
                    node=node,
                    context=self.context,
                    skip_manager=self.skip_manager,
                    loop_controller=self.loop_controller,
                    initial_inputs=initial_inputs if node['type'] == 'startNode' else None
                )
                
                self.context.set_node_output(node_id, output, cost)
                self.context.increment_execution_count(node_id)
                
                await self._send_update(node_id, "completed", output=output)
                
                # Handle loop completion
                if self.loop_controller.is_loop_node(node_id):
                    self.loop_controller.mark_iteration(node_id)
                    if self.loop_controller.should_restart_loop(node_id):
                        loop_start_nodes = self.loop_controller.get_loop_start_nodes(node_id)
                        execution_queue.extend(loop_start_nodes)
                        continue
                
                # Queue next nodes based on execution result
                next_nodes = self._get_next_nodes(node_id, output)
                execution_queue.extend(next_nodes)
                
            except Exception as e:
                error_msg = str(e)
                self.context.set_error(node_id, error_msg)
                await self._send_update(node_id, "error", error=error_msg)
                # Continue execution with other branches
        
        return self.context.node_outputs, self.context.total_cost
    
    def _get_start_nodes(self) -> List[str]:
        """Get all start nodes in the diagram.
        
        Returns:
            List of start node IDs
        """
        return [
            node_id
            for node_id, node in self.nodes.items()
            if node['type'] == 'startNode'
        ]
    
    def _should_skip_node(self, node: Dict[str, Any]) -> bool:
        """Check if a node should be skipped.
        
        Args:
            node: The node to check
            
        Returns:
            True if node should be skipped
        """
        node_id = node['id']
        execution_count = self.context.get_execution_count(node_id)
        max_iterations = node.get('data', {}).get('maxIterations')
        
        return self.skip_manager.should_skip(node_id, execution_count, max_iterations)
    
    def _get_next_nodes(self, node_id: str, output: Any = None) -> List[str]:
        """Get the next nodes to execute based on current node output.
        
        Args:
            node_id: Current node ID
            output: Output from the current node (used for conditions)
            
        Returns:
            List of next node IDs to execute
        """
        next_nodes = []
        outgoing = self.outgoing_arrows.get(node_id, [])
        
        for arrow in outgoing:
            # Handle conditional arrows
            if arrow.get('label') in ['true', 'false'] and isinstance(output, bool):
                if (arrow['label'] == 'true' and output) or (arrow['label'] == 'false' and not output):
                    next_nodes.append(arrow['target'])
            else:
                # Non-conditional arrow
                next_nodes.append(arrow['target'])
        
        return next_nodes
    
    async def _send_update(self, node_id: str, status: str, output: Any = None, error: str = None) -> None:
        """Send a streaming update if callback is configured.
        
        Args:
            node_id: The node ID
            status: The node status (running, completed, error, skipped)
            output: Optional output data
            error: Optional error message
        """
        if self.streaming_update_callback:
            update = {
                "node_id": node_id,
                "status": status,
                "timestamp": self.context.get_execution_time()
            }
            
            if output is not None:
                update["output"] = output
            
            if error:
                update["error"] = error
            
            await self.streaming_update_callback(update)
    
    def get_execution_summary(self) -> Dict[str, Any]:
        """Get a summary of the execution.
        
        Returns:
            Execution summary dictionary
        """
        return self.context.get_summary()

File: apps/web/src/engine/core/execution-engine.ts
/**
 * Core execution engine for orchestrating diagram execution.
 * 
 * This engine is environment-agnostic and can run in both browser and Node.js environments.
 * It uses dependency injection for executors, allowing different implementations for client vs server.
 */

import { 
  Diagram, 
  ExecutionOptions, 
  ExecutionResult, 
  ExecutionContext,
  ExecutionContext as TypedExecutionContext,
  ExecutionMetadata,
  ExecutionStatus,
  Node,
  NodeType,
  DiagramNode,
  DiagramArrow,
  ExecutionError,
  NodeExecutionError,
  DiagramExecutionError,
  MaxIterationsError,
  DEFAULT_MAX_ITERATIONS,
  DEFAULT_EXECUTION_TIMEOUT,
  BaseExecutorInterface,
  ExecutorResult
} from '@/shared/types/core';
import { createExecutionContext } from '../index';
import { DependencyResolver } from '../flow/dependency-resolver';
import { ExecutionPlanner } from '../flow/execution-planner';
import { SkipManager } from './skip-manager';
import { LoopController } from './loop-controller';

export interface ExecutorFactory {
  /**
   * Create an executor for a specific node type
   */
  createExecutor(nodeType: string): BaseExecutorInterface;
  
  /**
   * Check if a node type can be executed in the current environment
   */
  canExecute(nodeType: string): boolean;
  
  /**
   * Get all supported node types in this environment
   */
  getSupportedNodeTypes(): string[];
}

export interface StreamManager {
  /**
   * Emit a stream update
   */
  emit(update: any): void;
  
  /**
   * Check if streaming is enabled
   */
  isEnabled(): boolean;
}

export class ExecutionEngine {
  private dependencyResolver?: DependencyResolver;
  private executionPlanner?: ExecutionPlanner;
  private skipManager: SkipManager;
  private loopController?: LoopController;
  
  // Node type mappings between React Flow types and core execution types
  private static readonly NODE_TYPE_MAP: Record<string, string> = {
    'startNode': 'start',
    'personJobNode': 'person_job',
    'personBatchJobNode': 'person_batch_job',
    'conditionNode': 'condition',
    'dbNode': 'db',
    'jobNode': 'job',
    'endpointNode': 'endpoint'
  };
  
  constructor(
    private executorFactory: ExecutorFactory,
    private streamManager?: StreamManager
  ) {
    this.skipManager = new SkipManager();
  }

  /**
   * Normalize node types from React Flow types to core execution types
   */
  private normalizeNodeType(nodeType: string): string {
    return ExecutionEngine.NODE_TYPE_MAP[nodeType] || nodeType;
  }

  /**
   * Normalize diagram nodes to use core execution types
   */
  private normalizeDiagram(diagram: Diagram): Diagram {
    return {
      ...diagram,
      nodes: diagram.nodes.map(node => ({
        ...node,
        type: this.normalizeNodeType(node.type) as NodeType
      }))
    };
  }

  /**
   * Execute a diagram with the given options
   */
  async execute(diagram: Diagram, options: ExecutionOptions = {}): Promise<ExecutionResult> {
    const executionId = this.generateExecutionId();
    const startTime = Date.now();
    
    // Normalize node types before execution
    const normalizedDiagram = this.normalizeDiagram(diagram);
    
    // Initialize execution metadata
    const metadata: ExecutionMetadata = {
      executionId,
      startTime,
      totalCost: 0,
      nodeCount: normalizedDiagram.nodes.length,
      status: 'running'
    };

    // Emit execution started event
    this.emitStreamUpdate({
      type: 'execution_started',
      executionId,
      data: { diagram: normalizedDiagram, options },
      timestamp: new Date()
    });

    try {
      // Create execution context from normalized diagram
      const context = this.createTypedExecutionContext(normalizedDiagram, executionId, startTime);
      
      // Initialize components
      this.initializeComponents(normalizedDiagram, context, options);
      
      // Validate diagram before execution
      if (!options.skipValidation) {
        await this.validateDiagram(normalizedDiagram, context);
      }
      
      // Execute the diagram
      const finalOutputs = await this.executeNodes(normalizedDiagram, context, options);
      
      // Create successful result
      const result: ExecutionResult = {
        success: true,
        context,
        metadata: {
          ...metadata,
          endTime: Date.now(),
          status: 'completed',
          totalCost: context.totalCost
        },
        finalOutputs,
        errors: []
      };

      this.emitStreamUpdate({
        type: 'execution_completed',
        executionId,
        data: result,
        timestamp: new Date()
      });

      return result;

    } catch (error) {
      const executionError = this.handleExecutionError(error, executionId);
      
      const result: ExecutionResult = {
        success: false,
        context: this.createTypedExecutionContext(normalizedDiagram, executionId, startTime),
        metadata: {
          ...metadata,
          endTime: Date.now(),
          status: 'failed',
          totalCost: 0
        },
        finalOutputs: {},
        errors: [executionError]
      };

      this.emitStreamUpdate({
        type: 'execution_failed',
        executionId,
        data: { error: executionError },
        timestamp: new Date()
      });

      return result;
    }
  }

  /**
   * Initialize all execution components
   */
  private initializeComponents(
    diagram: Diagram, 
    context: TypedExecutionContext, 
    options: ExecutionOptions
  ): void {
    // Create legacy context for existing components
    const legacyContext = createExecutionContext(
      diagram.nodes.map(n => ({ ...n, data: n.data as any })) as DiagramNode[],
      diagram.arrows as DiagramArrow[]
    );
    
    this.dependencyResolver = new DependencyResolver(legacyContext);
    this.executionPlanner = new ExecutionPlanner(legacyContext, this.dependencyResolver);
    
    // Initialize loop controller if needed
    const cycles = this.dependencyResolver.detectCycles();
    if (cycles.length > 0) {
      const loopNodes = cycles.flat();
      this.loopController = new LoopController(
        options.maxIterations || DEFAULT_MAX_ITERATIONS,
        loopNodes
      );
    }
    
    // Clear skip manager for fresh execution
    this.skipManager.clear();
  }

  /**
   * Validate the diagram before execution
   */
  private async validateDiagram(diagram: Diagram, context: TypedExecutionContext): Promise<void> {
    if (!this.dependencyResolver) {
      throw new DiagramExecutionError('Dependency resolver not initialized');
    }

    // Check for cycles
    const cycles = this.dependencyResolver.detectCycles();
    if (cycles.length > 0) {
      console.warn(`Diagram contains ${cycles.length} cycle(s). Loop controller will manage execution.`);
    }

    // Validate start nodes
    const startNodes = this.dependencyResolver.validateStartNodes();
    if (startNodes.length === 0) {
      throw new DiagramExecutionError('No start nodes found in diagram');
    }

    // Validate node types are supported
    for (const node of diagram.nodes) {
      if (!this.executorFactory.canExecute(node.type)) {
        throw new DiagramExecutionError(
          `Unsupported node type: ${node.type}`,
          { nodeId: node.id, nodeType: node.type }
        );
      }
    }
  }

  /**
   * Execute all nodes in the diagram
   */
  private async executeNodes(
    diagram: Diagram,
    context: TypedExecutionContext, 
    options: ExecutionOptions
  ): Promise<Record<string, any>> {
    if (!this.dependencyResolver || !this.executionPlanner) {
      throw new DiagramExecutionError('Execution components not initialized');
    }

    const startTime = Date.now();
    const timeout = options.timeout || DEFAULT_EXECUTION_TIMEOUT;
    const maxIterations = options.maxIterations || DEFAULT_MAX_ITERATIONS;
    
    const executedNodes = new Set<string>();
    const pendingNodes = new Set<string>();
    let currentIteration = 0;

    // Get start nodes
    const startNodes = this.dependencyResolver.validateStartNodes();
    pendingNodes.clear();
    startNodes.forEach(nodeId => pendingNodes.add(nodeId));

    while (pendingNodes.size > 0 && currentIteration < maxIterations) {
      // Check timeout
      if (Date.now() - startTime > timeout) {
        throw new DiagramExecutionError('Execution timeout exceeded');
      }

      currentIteration++;
      const nodesToExecute = Array.from(pendingNodes);
      pendingNodes.clear();
      let nodesExecutedThisIteration = 0;

      // Execute nodes in this iteration
      for (const nodeId of nodesToExecute) {
        try {
          const currentNode = diagram.nodes.find(n => n.id === nodeId);
          console.log(`[DEBUG] Attempting to execute node ${nodeId} (type: ${currentNode?.type})`);
          
          // Check if node should be skipped
          if (this.skipManager.isSkipped(nodeId)) {
            console.log(`Skipping node ${nodeId}: ${this.skipManager.getSkipReason(nodeId)}`);
            continue;
          }

          // Check if node has reached max iterations
          if (currentNode) {
            const executionCount = context.nodeExecutionCounts[nodeId] || 0;
            const maxIterations = currentNode.data.iterationCount || currentNode.data.maxIterations;
            
            if (maxIterations && this.skipManager.shouldSkip(nodeId, executionCount, maxIterations)) {
              console.log(`Skipping node ${nodeId}: max iterations (${maxIterations}) reached`);
              
              // For skipped nodes, we still need to process their outgoing connections
              // This ensures the flow continues even when nodes are skipped
              // IMPORTANT: During skipping, the forget rule does not apply - all inputs are counted
              // regardless of whether they came from first-only or default handles
              const nextNodes = this.dependencyResolver.getNextNodes(nodeId, context.conditionValues);
              nextNodes.forEach(nextNodeId => {
                if (!executedNodes.has(nextNodeId) && !pendingNodes.has(nextNodeId)) {
                  pendingNodes.add(nextNodeId);
                }
              });
              
              // Mark the node as having been executed (for dependency checking)
              // even though it was skipped
              executedNodes.add(nodeId);
              
              continue;
            }
          }

          // Check dependencies
          const [dependenciesMet, validArrows] = this.dependencyResolver.checkDependenciesMet(
            nodeId, 
            executedNodes, 
            context.conditionValues
          );

          console.log(`[DEBUG] Dependencies met for ${nodeId}: ${dependenciesMet}, executed nodes: [${Array.from(executedNodes).join(', ')}]`);

          if (!dependenciesMet) {
            // Re-add to pending if dependencies not met
            console.log(`[DEBUG] Re-adding ${nodeId} to pending due to unmet dependencies`);
            pendingNodes.add(nodeId);
            continue;
          }

          // Check if this is a PersonJob or Job node executing with first-only for the first time
          if (currentNode && (currentNode.type === 'person_job' || currentNode.type === 'job')) {
            const hasFirstOnlyInput = validArrows.some(arrow => 
              arrow.targetHandle?.endsWith('-input-first') || arrow.data?.handleMode === 'first_only'
            );
            
            if (hasFirstOnlyInput && !context.firstOnlyConsumed[nodeId]) {
              // Mark first-only as consumed for this node
              context.firstOnlyConsumed[nodeId] = true;
            }
          }

          // Execute the node
          await this.executeNode(nodeId, diagram, context, options);
          executedNodes.add(nodeId);
          nodesExecutedThisIteration++;
          console.log(`[DEBUG] Node ${nodeId} executed successfully`);

          // Get next nodes to execute
          const nextNodes = this.dependencyResolver.getNextNodes(nodeId, context.conditionValues);
          console.log(`[DEBUG] Next nodes from ${nodeId}: [${nextNodes.join(', ')}]`);
          nextNodes.forEach(nextNodeId => {
            if (!executedNodes.has(nextNodeId) && !pendingNodes.has(nextNodeId)) {
              pendingNodes.add(nextNodeId);
              console.log(`[DEBUG] Added ${nextNodeId} to pending nodes`);
            }
          });

        } catch (error) {
          // Handle node execution error
          const nodeError = error instanceof NodeExecutionError 
            ? error 
            : new NodeExecutionError(
                `Error executing node ${nodeId}: ${error instanceof Error ? error.message : String(error)}`,
                nodeId,
                diagram.nodes.find(n => n.id === nodeId)?.type || 'start',
                { originalError: error }
              );

          context.errors[nodeId] = nodeError.message;
          
          // Decide whether to continue or stop execution
          if (options.debugMode) {
            console.error(`Node execution failed: ${nodeError.message}`, nodeError);
            // In debug mode, continue with other nodes
            executedNodes.add(nodeId); // Mark as executed to prevent retry
          } else {
            throw nodeError;
          }
        }
      }

      // If no progress was made and we still have pending nodes, there might be a dependency issue
      if (pendingNodes.size > 0 && nodesExecutedThisIteration === 0) {
        const pendingArray = Array.from(pendingNodes);
        throw new DiagramExecutionError(
          `Execution deadlock detected. Pending nodes: ${pendingArray.join(', ')}`,
          { pendingNodes: pendingArray, currentIteration }
        );
      }
    }

    if (currentIteration >= maxIterations) {
      throw new MaxIterationsError(
        `Maximum iterations (${maxIterations}) exceeded`,
        { currentIteration, pendingNodes: Array.from(pendingNodes) }
      );
    }

    return context.nodeOutputs;
  }

  /**
   * Execute a single node
   */
  private async executeNode(
    nodeId: string,
    diagram: Diagram,
    context: TypedExecutionContext,
    options: ExecutionOptions
  ): Promise<void> {
    const node = diagram.nodes.find(n => n.id === nodeId);
    if (!node) {
      throw new NodeExecutionError(
        `Node ${nodeId} not found in diagram`,
        nodeId,
        'start' as any
      );
    }

    // Emit node started event
    this.emitStreamUpdate({
      type: 'node_started',
      executionId: context.executionId,
      nodeId,
      data: { node },
      timestamp: new Date()
    });

    try {
      // Get executor for this node type
      const executor = this.executorFactory.createExecutor(node.type);
      
      // Validate inputs
      const validation = await executor.validateInputs(node, context);
      if (!validation.isValid) {
        throw new NodeExecutionError(
          `Input validation failed: ${validation.errors.join(', ')}`,
          nodeId,
          node.type,
          { validationErrors: validation.errors }
        );
      }

      // Execute the node
      const result = await executor.execute(node, context, options);
      
      // Update context with results
      context.nodeOutputs[nodeId] = result.output;
      context.totalCost += result.cost;
      context.nodeExecutionCounts[nodeId] = (context.nodeExecutionCounts[nodeId] || 0) + 1;
      context.executionOrder.push(nodeId);

      // Handle condition node results
      if (node.type === 'condition' && typeof result.output === 'boolean') {
        context.conditionValues[nodeId] = result.output;
      }

      // Emit node completed event
      this.emitStreamUpdate({
        type: 'node_completed',
        executionId: context.executionId,
        nodeId,
        data: { result },
        timestamp: new Date()
      });

    } catch (error) {
      // Emit node failed event
      this.emitStreamUpdate({
        type: 'node_failed',
        executionId: context.executionId,
        nodeId,
        data: { error: error instanceof Error ? error.message : String(error) },
        timestamp: new Date()
      });

      throw error;
    }
  }

  /**
   * Create a typed execution context from a diagram
   */
  private createTypedExecutionContext(
    diagram: Diagram, 
    executionId: string, 
    startTime: number
  ): TypedExecutionContext {
    // Create nodes by ID map
    const nodesById: Record<string, any> = {};
    const incomingArrows: Record<string, any[]> = {};
    const outgoingArrows: Record<string, any[]> = {};

    // Initialize maps
    for (const node of diagram.nodes) {
      nodesById[node.id] = node;
      incomingArrows[node.id] = [];
      outgoingArrows[node.id] = [];
    }

    // Populate arrow maps
    for (const arrow of diagram.arrows) {
      if (arrow.source && arrow.target) {
        const sourceArrows = outgoingArrows[arrow.source] || [];
        const targetArrows = incomingArrows[arrow.target] || [];
        
        outgoingArrows[arrow.source] = [...sourceArrows, arrow];
        incomingArrows[arrow.target] = [...targetArrows, arrow];
      }
    }

    return {
      executionId,
      nodeOutputs: {},
      nodeExecutionCounts: {},
      totalCost: 0,
      startTime,
      errors: {},
      executionOrder: [],
      conditionValues: {},
      firstOnlyConsumed: {},
      diagram: null,
      nodesById,
      outgoingArrows,
      incomingArrows
    };
  }

  /**
   * Handle and standardize execution errors
   */
  private handleExecutionError(error: any, executionId: string): ExecutionError {
    if (error instanceof NodeExecutionError || error instanceof DiagramExecutionError) {
      return {
        nodeId: error instanceof NodeExecutionError ? error.nodeId : undefined,
        nodeType: error instanceof NodeExecutionError ? error.nodeType : undefined,
        message: error.message,
        details: error.details,
        timestamp: new Date(),
        stack: error.stack
      };
    }

    return {
      message: error.message || 'Unknown execution error',
      details: { originalError: error },
      timestamp: new Date(),
      stack: error.stack
    };
  }

  /**
   * Emit a stream update if streaming is enabled
   */
  private emitStreamUpdate(update: any): void {
    if (this.streamManager && this.streamManager.isEnabled()) {
      this.streamManager.emit(update);
    }
  }

  /**
   * Generate a unique execution ID
   */
  private generateExecutionId(): string {
    return `exec_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
  }

  /**
   * Get the current execution status
   */
  getStatus(): ExecutionStatus {
    // This could be enhanced to track actual execution state
    return 'pending';
  }

  /**
   * Stop the current execution
   */
  async stop(): Promise<void> {
    // Implementation for stopping execution
    // This could involve setting a stop flag that is checked during execution loops
    console.log('Execution stop requested');
  }
}